[SPARK-11617] [network] Fix leak in TransportFrameDecoder.

vanzin opened the pull request:
The code was using the wrong API to add data to the internal composite
buffer, causing buffers to leak in certain situations. Use the right
API and enhance the tests to catch memory leaks.

Also, avoid reusing the composite buffers when downstream handlers keep
references to them; this seems to cause a few different issues even though
the ref counting code seems to be correct, so instead pay the cost of copying
a few bytes when that situation happens.

Commit:
[SPARK-11617] [network] Fix leak in TransportFrameDecoder.

The code was using the wrong API to add data to the internal composite
buffer, causing buffers to leak in certain situations. Use the right
API and enhance the tests to catch memory leaks.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45603 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45603/consoleFull)** for PR 9619 at commit [`ed0c1d7`](https://github.com/apache/spark/commit/ed0c1d7e6df3357344574b3ef2dccb526ed3f9b7).

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45608 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45608/consoleFull)** for PR 9619 at commit [`ed0c1d7`](https://github.com/apache/spark/commit/ed0c1d7e6df3357344574b3ef2dccb526ed3f9b7).

SparkQA commented:
**[Test build #45603 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45603/consoleFull)** for PR 9619 at commit [`ed0c1d7`](https://github.com/apache/spark/commit/ed0c1d7e6df3357344574b3ef2dccb526ed3f9b7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class JavaLBFGSExample `\n  * `class LDA @Since(\"1.6.0\") (`\n  * `  case class Metadata(`\n  * `      require(className == expectedClassName, s\"Error loading metadata: Expected class name\" +`\n  * `class SlidingRDDPartition[T](val idx: Int, val prev: Partition, val tail: Seq[T], val offset: Int)`\n  * `class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int, val step: Int)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45603/
Test FAILed.

SparkQA commented:
**[Test build #45608 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45608/consoleFull)** for PR 9619 at commit [`ed0c1d7`](https://github.com/apache/spark/commit/ed0c1d7e6df3357344574b3ef2dccb526ed3f9b7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45608/
Test PASSed.

Commit:
Copy bytes when handlers keep references to data.

This makes the frame decoder behave more like netty's ByteToMessageDecoder,
at the expense of copying some data in a few cases.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45779/consoleFull)** for PR 9619 at commit [`8a7d194`](https://github.com/apache/spark/commit/8a7d19496c8c61f0e3e2c5180857f1773ba10372).

SparkQA commented:
**[Test build #45779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45779/consoleFull)** for PR 9619 at commit [`8a7d194`](https://github.com/apache/spark/commit/8a7d19496c8c61f0e3e2c5180857f1773ba10372).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45779/
Test PASSed.

vanzin commented:
/cc @zsxwing @rxin

Commit:
Remove stray import.

SparkQA commented:
**[Test build #45909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45909/consoleFull)** for PR 9619 at commit [`180456c`](https://github.com/apache/spark/commit/180456ce3d25a126a74fa0f57c22665ef002b6cd).

SparkQA commented:
**[Test build #45909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45909/consoleFull)** for PR 9619 at commit [`180456c`](https://github.com/apache/spark/commit/180456ce3d25a126a74fa0f57c22665ef002b6cd).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45909/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #45999 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45999/consoleFull)** for PR 9619 at commit [`180456c`](https://github.com/apache/spark/commit/180456ce3d25a126a74fa0f57c22665ef002b6cd).

SparkQA commented:
**[Test build #45999 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45999/consoleFull)** for PR 9619 at commit [`180456c`](https://github.com/apache/spark/commit/180456ce3d25a126a74fa0f57c22665ef002b6cd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45999/
Test PASSed.

Commit:
Better test for retained frames.

This test actually fails with java.lang.IndexOutOfBoundsException if the
fix in this patch set is disabled.

Commit:
Refactor some common test code.

zsxwing commented on network/common/src/main/java/org/apache/spark/network/util/TransportFrameDecoder.java:
```diff
@@ -56,32 +56,43 @@ public void channelRead(ChannelHandlerContext ctx, Object data) throws Exception
+  private void discardReadBytes() {
+    // If the buffer's been retained by downstream code, then make a copy of the remaining
+    // bytes into a new buffer. Otherwise, just discard stale components.
+    if (buffer.refCnt() > 1) {
```
@vanzin could you give a real case? Or this is just for correctness, even if downstream in Spark doesn't use `retain`?

vanzin replied:
Actually Spark does use retain() when fetching shuffle blocks, and for some reason that causes problems. I think the real problem is somewhere in netty code, but this is the workaround the netty code itself uses (see ByteToMessageDecoder).

zsxwing replied:
Yeah. Just saw `retain()` in `ChunkFetchSuccess`.

SparkQA commented:
**[Test build #46019 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46019/consoleFull)** for PR 9619 at commit [`7fe9617`](https://github.com/apache/spark/commit/7fe96174edde015c8c3bc43ca734e809c5041192).

zsxwing commented:
LGTM pending tests.

SparkQA commented:
**[Test build #46015 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46015/consoleFull)** for PR 9619 at commit [`dcdfc31`](https://github.com/apache/spark/commit/dcdfc31bfa7461c16a5a3cd17ae66cc999697664).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46015/
Test FAILed.

zsxwing commented:
retest this please

zsxwing commented:
> Test FAILed.
> Refer to this link for build results (access rights to CI server needed): 
> https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46015/
> Test FAILed.

This is a bug that will be fixed in #9707

SparkQA commented:
**[Test build #46024 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46024/consoleFull)** for PR 9619 at commit [`7fe9617`](https://github.com/apache/spark/commit/7fe96174edde015c8c3bc43ca734e809c5041192).

SparkQA commented:
**[Test build #46019 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46019/consoleFull)** for PR 9619 at commit [`7fe9617`](https://github.com/apache/spark/commit/7fe96174edde015c8c3bc43ca734e809c5041192).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46019/
Test FAILed.

SparkQA commented:
**[Test build #46024 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46024/consoleFull)** for PR 9619 at commit [`7fe9617`](https://github.com/apache/spark/commit/7fe96174edde015c8c3bc43ca734e809c5041192).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46024/
Test PASSed.

vanzin commented:
Merging this (master / 1.6).
